hooks/PlayerControl.cpp

```diff
@@ -334,6 +334,17 @@ void dGameObject_SetActive(GameObject* __this, bool value, MethodInfo* method)
 			}
 		}
 	}
+	// Right-click Teleport
+	if (IsInGame() && State.RightClickTeleport && ImGui::IsMouseClicked(ImGuiMouseButton_Right)) {
```

Comment:
We need to find a better place for this functionality. @OsOmE1 or @cddjr any ideas ?

Reply:
i'm confused. _SeeProtect() isn't a loop, then it uses the hook of dPlayerControl_TurnOnProtection()

how can _SeeProtect() runs in a loop with just one call?

Reply:
here's what I did so far, obviously it doesn't work

![image](https://user-images.githubusercontent.com/51813377/207587670-12ea993d-6d80-40ac-8087-09414fc0e006.png)

Reply:
> Yeah I don't like how `__SeeProtect()` is implemented. It'd only work when the menu itself is being rendered. I think both of them should be moved to the `dPlayerControl_FixedUpdate` hook. This runs every 12ish milliseconds and it's in a good gamestate so RPCs should work.

i tried putting it on fixed update and it was hella janky

Reply:
> You'd only want to run it for the localplayer. So put it inside the block that checks `if _this == *Game::pLocalPlayer`. Alternatively (or in addition to), add a timer to prevent the click input from being processed multiple times.

it was still janky even without timer. should i add another hook that just a regular update? (not fixed update)

![image](https://user-images.githubusercontent.com/51813377/207591131-ff2a6b58-c4aa-4581-a76b-b35afc7d6e9f.png)

Reply:
@kotae4 we might want to expand the framework and add functionality to add Taskrunners which run on their own in loop. This way we can move all the functionality which need to run in loops into their own runners without cluttering all the hooked functions and potentially tanking the fps

Reply:
@GDjkhp Good point, forgot the golden rule of "never check input in FixedUpdate". Yeah, either hook PlayerControl::Update (if it has one) or use the existing `dInnerNetClient_Update` hook.

@v0idp Well it wouldn't improve the FPS since we'd have to invoke all of those from some hooked Update function, but it would be nice to have all of that general 'run-every-frame' logic consolidated into one place.

Reply:
@kotae4 it really depends if we invoke anything in the hooked function or if we just use it to run stuff on frame which we can honestly avoid with our own taskrunners. I imagine that taskrunners can be defined for each featureset like self, game, etc. and then we can do all the loopy stuff for those categories in their own respective function loop or in this case taskrunner function

Reply:
Setting the Position shouldn't need to be run from a virtual thread right?
if you use UnityEngine.Transform::set_localPosition_Injected(UnityEngine.Vector3&)
you should just be able to do it in a callback for like OnRightMouseButton. 
Otherwise, I think it might be good to use the Unity input system no?

Reply:
that's only clientside and won't work. you will need to issue an rpc event to make it happen and also apply to the whole lobby

Reply:
Ahh okay I didnt think about that